packages/extension/src/ui/features/ledger/utils.ts

```diff
@@ -24,8 +26,35 @@ export const getLedgerApp = async () => {
   return app
 }
 
-export const deriveAccount = async (targetAddressGroup?: number) => {
+const isAddressGroupMatched = (account: WalletAccount, targetAddressGroup: number | undefined) => {
+  if (targetAddressGroup === undefined) {
+    return true
+  } else {
+    return account.signer.group === targetAddressGroup
+  }
+}
+
+export const deriveAccount = async (existingLedgerAccounts: WalletAccount[], targetAddressGroup: number | undefined, keyType: string) => {
+  if (keyType !== "default" && keyType !== "bip340-schnorr") {
+    throw Error("Unsupported key type: " + keyType)
+  }
+
   const app = await getLedgerApp()
-  const path = `m/44'/1234'/0'/0/0`
-  return await app.getAccount(path, targetAddressGroup)
+  const relevantAccounts = existingLedgerAccounts.filter((account) =>
+    account.signer.keyType === keyType && isAddressGroupMatched(account, targetAddressGroup)
+  )
+  relevantAccounts.sort((a, b) => a.signer.derivationIndex - b.signer.derivationIndex)
+  console.log(`===== relevantAccounts`, relevantAccounts)
+  for (let i = 0; i < relevantAccounts.length; i++) {
+    const existingAccount = relevantAccounts[i]
+    const path = getHDWalletPath(keyType, existingAccount.signer.derivationIndex + 1)
+    console.log(`===== path`, path)
+    const [newAccount, hdIndex] = await app.getAccount(path, targetAddressGroup, keyType)
+    if (existingLedgerAccounts.find((account) => account.address === newAccount.address) === undefined) {
+      return [newAccount, hdIndex] as const
+    }
+  }
+  const path = getHDWalletPath(keyType, 0)
```

Comment:
Q: with the for loop above, is it possible that we end up here?

Reply:
Yes, it is possible because the number of relevant accounts might be 0